src/csharp/Grpc.Core.Api/DeserializationContext.cs

```diff
@@ -48,6 +96,23 @@ public virtual byte[] PayloadAsNewBuffer()
             throw new NotImplementedException();
         }
 
+        /// <summary>
+        /// Gets the entire payload as a leased byte array.
+        /// The caller is expected to call Dispose exactly once, returning the buffer to the pool. Accessing
+        /// the data after disposal or calling Dispose multiple times will result in undefined behavior.
+        /// Only the range defined by Offset and Count shoud be accessed by the consumer.
+        /// </summary>
+        /// <returns>a leased buffer containing the entire payload</returns>
+        public virtual LeasedBuffer PayloadAsLeasedBuffer()
```

Comment:
I've experimented with similar concept in the past ( PayloadAsRentedBuffer() ) which was using ArrayPool.Shared.Rent internally, but I decided not to add that method as I wasn't sure it was providing much extra value.
Once you're on netstandard2.0 (= GRPC_CSHARP_SUPPORT_SYSTEM_MEMORY is true), the can provide a deserializer that can easily perform pooling (custom or using ArrayPool) and read the content from PayloadAsReadOnlySequence().    This doesn't help .NET desktop users, but my assumption was is that most users that would care about performance would already be using .NET core.

Reply:
I don't think that final assumption is valid: many folks care deeply but are still attempting gradual migration of large code-bases. But it may be moot - see other comment